internal/engine/synclet_build_and_deployer.go

```diff
@@ -26,36 +26,59 @@ type SyncletBuildAndDeployer struct {
 
 	kCli k8s.Client
 
-	deployInfo   map[docker.ImgNameAndTag]DeployInfo
+	deployInfo   map[docker.ImgNameAndTag]*DeployInfo
 	deployInfoMu sync.Mutex
 }
 
 type DeployInfo struct {
 	podID       k8s.PodID
 	containerID k8s.ContainerID
 	nodeID      k8s.NodeID
+
+	ready chan struct{} // Close this channel when the DeployInfo is populated
+}
+
+// ~~~ should these verify that the channel is non-nil?
+func (di *DeployInfo) waitUntilReady() { <-di.ready }
+func (di *DeployInfo) markReady()      { close(di.ready) }
+
+func newEmptyDeployInfo() *DeployInfo {
+	return &DeployInfo{ready: make(chan struct{})}
 }
 
 func NewSyncletBuildAndDeployer(kCli k8s.Client, ssm SidecarSyncletManager) *SyncletBuildAndDeployer {
 	return &SyncletBuildAndDeployer{
 		kCli:       kCli,
-		deployInfo: make(map[docker.ImgNameAndTag]DeployInfo),
+		deployInfo: make(map[docker.ImgNameAndTag]*DeployInfo),
 		ssm:        ssm,
 	}
 }
 
-func (sbd *SyncletBuildAndDeployer) getDeployInfoForImage(img reference.NamedTagged) (DeployInfo, bool) {
+func (sbd *SyncletBuildAndDeployer) deployInfoForImage(img reference.NamedTagged) (*DeployInfo, bool) {
 	sbd.deployInfoMu.Lock()
+	defer sbd.deployInfoMu.Unlock()
 	deployInfo, ok := sbd.deployInfo[docker.ToImgNameAndTag(img)]
-	sbd.deployInfoMu.Unlock()
 	return deployInfo, ok
 }
 
-func (sbd *SyncletBuildAndDeployer) setDeployInfoForImage(img reference.NamedTagged, deployInfo DeployInfo) {
-	sbd.deployInfoMu.Lock()
-	key := docker.ToImgNameAndTag(img)
-	sbd.deployInfo[key] = deployInfo
-	sbd.deployInfoMu.Unlock()
+func (sbd *SyncletBuildAndDeployer) deployInfoForImageOrNew(img reference.NamedTagged) (*DeployInfo, bool) {
+	deployInfo, ok := sbd.deployInfoForImage(img)
```

Comment:
let's inline the map access, and make a single Lock()/Unlock() surround the entire function. right now this has race conditions